### ISSUE TITLE
Mutation of data in onSuccess callback

### ISSUE THREAD

**Issue by SchalkBezuidenhout** [vercel/swr#2332]
# Bug report

## Description / Observed Behavior

What kind of issues did you encounter with SWR?
In v.1.3.0, I was able to mutate `data `in the `onSuccess `callback, e.g. 
`onSuccess: (data) => mutate(data.map(item => ({id: item.id, description: item.description})), false)`

In v2.0.0, `data` and the cache doesn't get mutated in the `onSuccess` callback and `data` remains the raw data received from the API.

## Expected Behavior

How did you expect SWR to behave here?
Like in v.1.3.0 where I could mutate data received from the same API for different use cases.

## Repro Steps / Code Example

```
const sortOrderFormat = (data) => {
    return data?.meta?.map(item => {
        const details = {}
        const sanitize = string => DOMPurify.sanitize(string, { ALLOWED_TAGS: ['br'], IN_PLACE: true })
        const serviceAreas = item.description.match(/Service Areas:.+(?=Physical Address)/g)
        const rest = item.description.replace(serviceAreas, '')
        sanitize(rest)
            .replaceAll('&nbsp;', ' ')
            .replaceAll('&amp;', '&')
            .split('<br>')
            .forEach((entry, index) => {
                const kv = entry.split(':')
                const key = kv[0].trim()
                const value = kv[1]?.trim()
                switch (key) {
                    case 'E-mail':
                        const emailAddressRegex = ...
                        const emailAddress = value.match(emailAddressRegex)
                        return details[key] = emailAddress?.[0]
                        break;
                    case 'Magisterial District':
                        details[key] = value
                        return details['Service Areas'] = DOMPurify.sanitize(serviceAreas, { ALLOWED_TAGS: [], IN_PLACE: true }).split(':')?.[1]
                            ?.replaceAll('&nbsp;', ' ')
                            ?.replaceAll('&amp;', '&')
                            ?.split('.')
                            ?.join(', ')
                        break;
                    default:
                        return details[key] = value
                        break;
                }
            })

        return {
            name: item.title,
            address: item.address,
            details: details
        }
    })
}

const { data, mutate, isValidating } = useSWRImmutable('/api/sheriffs', {
        onSuccess: (data) => mutate(sortOrderFormat(data), false)
    })
```

in v.1.3.0, `data `will be the returned result from `sortOrderFormat(data)`, but not in v.2.0.0.

## Additional Context

SWR version. 2.0.0

**Comment by HarryNiotis** [comment]
I am actually seeing a similar behaviour using the global `mutate` while trying to upgrade to V2.
Example usage
```
const { mutate } = useSWRConfig()
const key = '123'

const { data } = useSWR(key, () => fetchStuff())

...

const someDeleteMethod = () => {

...do delete...

await mutate(key, null, false)

...
}
```
in `v1.1.0` that would update the cache and the `useSWR` hook will return `null` in the next re-render
in `v2.0.0` the cache is not updated and the `useSWR` hook still returns `data` from the cache

I did some debugging by getting the result of the `mutate` call back to check the cache in `v2.0.0`
```
const newData = await mutate(key, null, false)
console.log({ newData }) ---> newData: null
```
It seems that the mutate call actually works but somehow the hook still returns the old data.
I also tried
```
await mutate(key, null, { revalidate: false })
```

Trying the same scenario using the bound `mutate` works as expected
```
const key = '123'

const { data, mutate } = useSWR(key, () => fetchStuff())

...

const someDeleteMethod = () => {

...do delete...

await mutate(null, false) --> works
```

For context:
This is just a simplified example. We use this pattern to invalidate relevant resources using their keys when we do CRUD operations, so we use `mutate` with the appropriate keys. Is not necessarily to always `mutate` the `data` in the same hook
They `key` is always a static string in this instance and not a function, so it should not be affected by the changes introduced in `v2.0.0`

**Comment by SchalkBezuidenhout** [comment]
@HarryNiotis, this is quite concerning, because my global mutate seems to work fine, but the bound mutate in some instances (as in my example) does not, and in other instances it does.

I wonder if it has something to do with the format of the `key` or if something broke with the change implemented in `v2.0.0`
